File: src/main/java/uk/co/rafearnold/randomworkoutroutine/model/entity/Group.java
package uk.co.rafearnold.randomworkoutroutine.model.entity;

import com.fasterxml.jackson.annotation.JsonIgnore;
import lombok.Data;
import lombok.EqualsAndHashCode;

import javax.persistence.Entity;
import javax.persistence.FetchType;
import javax.persistence.ManyToMany;
import javax.persistence.Transient;
import java.util.List;

@Entity(name = "grouping")
@Data
@EqualsAndHashCode(callSuper = true)
public class Group extends Item {

    @ManyToMany(fetch = FetchType.EAGER)
    private List<ExerciseOption> exerciseOptions;

    @Transient
    @JsonIgnore
    private double[] optionWeights;
}

File: web/src/main/kotlin/uk/co/rafearnold/randomworkoutroutine/web/model/entity/Group.kt
package uk.co.rafearnold.randomworkoutroutine.web.model.entity

import com.fasterxml.jackson.annotation.JsonIgnore
import uk.co.rafearnold.randomworkoutroutine.model.Group
import java.util.*
import javax.persistence.*

/**
 * An implementation of [Group] for JPA.
 */
@Entity(name = "grouping")
class Group(
    @Id
    override var id: UUID = UUID.randomUUID(),
    @OneToMany(fetch = FetchType.EAGER)
    @OrderColumn
    override var exercises: MutableList<ExerciseOption> = mutableListOf(),
    @ManyToOne(cascade = [CascadeType.ALL])
    @JsonIgnore
    var routine: Routine? = null
) : Group {

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is Group) return false

        if (id != other.id) return false
        if (exercises != other.exercises) return false

        return true
    }

    override fun hashCode(): Int {
        var result = id.hashCode()
        result = 31 * result + exercises.hashCode()
        return result
    }
}
